### DISCUSSION TITLE
Decimal module wrapper for relativity._add_deflection function to attain greater precision?

### DISCUSSION THREAD

**Discussion by IvanKrasnyj** [skyfielders/python-skyfield#863] [Q&A]
In my research, the challenge is to improve the precision of relativistic gravitational deflection, which exceeds the standard capabilities of floats in Python. I am interested in such quantities as deflection of the position of Saturn by Moon near their conjunction on March 29, 2007 at 05:00, or deflection of the position of Mars by Moon near their conjunction on May 10, 2008 at 14:00. At the indicated time points, in particular, paroxysms of volcano Etna were registered.

In the model I am considering, extremely small gravitational deflections, caused, in particular, by the Moon, control the direction of large force vectors, which as a result induces significant gravitational perturbations of the observer. I only need more accurate calculations for the __add_deflection_ function, without corrections for aberration (light_time).

Do I understand correctly that I will be able to connect the _decimal_ module wrapper for the __add_deflection_ function, as well as the imported _dots_ and _length_of_ functions, in order to increase precision of position corrections for relativistic deflection, as well as for angles (RA, DEC)?

**Comment by brandon-rhodes** [comment]
Alas, no — for example, the `dots()` routine, if you examine its source code, uses a NumPy method to perform its `sum()`:

    return (v * u).sum(axis=0)

So Python `Decimal` objects cannot be passed to this function, because they lack a `sum()` method. If you want `Decimal` level precision, I would suggest rewriting the deflection routine using your own code, so that you can pass decimals in and get them back out as a result. Skyfield's own code can only serve as an example you copy, but won't be able to digest decimals itself.

**Reply by IvanKrasnyj** [reply]
Thanks a lot, Brandon. As far as I understand, there are no other alternatives to improve precision. I will try to reproduce functionality of the `_add_deflection` routine using the `decimal` module.

**Reply by IvanKrasnyj** [reply]
Besides `decimal` and `mpmath` there is more suitable option  `numpy.longouble.` Here is what I've got as a result for greater precision of the `_add_deflection` routine. `dots` and `length_of` routines also accept `numpy.longdouble`. 

```
# casting floats as numpy.longdouble (AKA numpy.longfloat or numpy.float128 in Linux x86_64)

from numpy import array, longdouble as ld

fac1L = ld(2.0 * GS / (C * C * emag * AU_M * rmass))
fac2L = ld(1.0 + qdote)
mul = fac1L / fac2L * ld(pmag)
ehatL = [ld(i) for i in ehat]
qhatL = [ld(i) for i in qhat]
pdotqL = ld(pdotq)
edotpL = ld(edotp)
pdotq_ehat = [pdotqL * i for i in ehatL]
edotp_qhat = [edotpL * i for i in qhatL]

position_update = array([((pdotq_ehat[i] - edotp_qhat[i]) * mul) for i in range(3)])

```

**Reply by brandon-rhodes** [reply]
Interesting! The last time I tried `longdouble` I think I must have been on a platform whose CPU didn't support it, because the result that came out of the whole computation was the same as without `longdouble`.

Note that the Saturn and Moon positions are specified by mere 64-bit floats in the NASA ephemerides, so unless you can find an alternative way of generating their positions — which would also include upgrading Skyfield's whole position computation apparatus to 128-bit floats — then all the 128-bit floats are going to do is extend the "noise" down past the final bits of the position and angle vectors; 128-bit floats won't magically "know" the position of Saturn any better than is described in the 64-bit floats with which NASA represents them.

I recommend consulting the documentation around the ephemeris you are using — its physical error/uncertainly for Mars and Saturn positions is probably bigger than the limit imposed by 64 bits in any case, so even the 64-bit floats have noise at the bottom bits, not real position data.

**Reply by IvanKrasnyj** [reply]
Of course, it would be useful to have high-precision ephemeris, but in our case, we can be content with what we have. I am engaged in the study of gravitational perturbations on the geosphere caused by the conjunction and opposition of planets (deflectors) relative to each other and to compact masses of deep space.

The absolute accuracy of the ephemeris has little effect on the geometry of the planetary alignment and the physics of the process, i.e. on the value of the relativistic gravitational deflection, which interests me. The inaccuracy of the ephemeris can only lead to an insignificant shift in time.

As written in the USGS FAQ, _"The stresses induced in the earth by an extraterrestrial mass are proportional to the gravitational field gradient dg( r ) / dr and NOT to the gravitational field g( r )."_ The classical theory of gravity does not find reasons for the occurrence of gradients of the gravitational field of an earthly observer caused by motion of the planets, however, we believe ( [The novel pushing gravity model and volcanic activity. Is alignment of planets with compact stars a possible cause of natural phenomena?](https://iopscience.iop.org/article/10.1088/1742-6596/2081/1/012019/pdf) ) that such transient perturbations occur and are controlled by a relativistic gravitational deflection due to movement of the planets.
